File: trading/src/test/kotlin/io/squer/devenv/containers/Redis.java
package io.squer.devenv.containers;

import java.util.Map;
import org.testcontainers.Testcontainers;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.Network;
import org.testcontainers.containers.wait.strategy.Wait;
import org.testcontainers.utility.DockerImageName;

public class Redis extends GenericContainer<Redis> {

  private static final String IMAGE_NAME = "docker.io/bitnami/redis:7.0";
  private static final int EXPOSED_PORT = 6379;
  private static final int HOST_PORT = 6379;
  private static final int CONTAINER_PORT = 6379;
  private static final Map<String, String> ENV = Map.of(
    "ALLOW_EMPTY_PASSWORD", "yes"
  );

  private Redis() {
    super(DockerImageName.parse(IMAGE_NAME));
  }

  private static GenericContainer<Redis> instance;

  public static GenericContainer<Redis> getInstance(final Network network) {
    if (instance == null) {
      instance = new Redis()
        .withEnv(ENV)
        .withNetwork(network)
        .withNetworkMode("bridge")
        .withExposedPorts(EXPOSED_PORT)
        .withCreateContainerCmdModifier(cmd -> cmd.withHostName("redis").withName("redis"))
        .waitingFor(
          Wait.forLogMessage(".*Ready to accept connections.*\\n", 1)
        );

      ((Redis) instance).configurePorts();
    }
    return instance;
  }

  private void configurePorts() {
    super.addFixedExposedPort(HOST_PORT, CONTAINER_PORT);
  }

}

File: trading/src/test/kotlin/io/squer/devenv/containers/Redis.kt
package io.squer.devenv.containers

import com.github.dockerjava.api.command.CreateContainerCmd
import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.Network
import org.testcontainers.containers.wait.strategy.Wait
import org.testcontainers.utility.DockerImageName

class Redis(network: Network?) : GenericContainer<Redis?>(DockerImageName.parse(IMAGE_NAME)) {
    init {
        addFixedExposedPort(HOST_PORT, CONTAINER_PORT)
        withEnv(ENV)
        withNetwork(network)
        withNetworkMode("bridge")
        withCreateContainerCmdModifier { cmd: CreateContainerCmd -> cmd.withHostName("redis").withName("redis") }
        waitingFor(
                Wait.forLogMessage(".*Ready to accept connections.*\\n", 1)
        )
    }

    companion object {
        private const val IMAGE_NAME = "docker.io/bitnami/redis:7.0"
        private const val HOST_PORT = 6379
        private const val CONTAINER_PORT = 6379
        private val ENV = mapOf("ALLOW_EMPTY_PASSWORD" to "yes")
    }
}